src/Core/Animation/DualQuaternionSkinning.hpp

```diff
@@ -9,29 +10,55 @@ namespace Ra {
 namespace Core {
 namespace Animation {
 
+struct SkinningRefData;
+struct SkinningFrameData;
+
 using DQList = AlignedStdVector<DualQuaternion>;
 
-/*
- * computeDQ computes the dual quaternions from a given pose and a given set of skinning weights.
+/// \name Dual Quaternion Skinning
+/// \{
+
+// clang-format off
+/**
+ * \brief Computes the per-vertex dual quaternion.
  *
- * WARNING : in Debug the function will assert if pose and weight size mismatch. In Release will
- * simply crash.
+ * Such a dual quaternion is computed as:
+ * \f$ \mathbf{Q}_i=\frac{\sum_{s\in S} \omega_{is}\mathbf{Q}_s}
+ *                       {\|\sum_{s\in S} \omega_{is}\mathbf{Q}_s\|} \f$
  */
-void RA_CORE_API computeDQ( const Pose& pose, const Sparse& weight, DQList& DQ );
+// clang-format on
+DQList RA_CORE_API computeDQ( const Pose& pose, const WeightMatrix& weight );
 
-// Same version, without the parallelism for reference purposes (see github issue #118)
-void RA_CORE_API computeDQ_naive( const Pose& pose, const Sparse& weight, DQList& DQ );
+/**
+ * \brief Default non-optimized, non-parallel implementation of computeDQ.
+ */
+DQList RA_CORE_API computeDQ_naive( const Pose& pose, const WeightMatrix& weight );
 
-/*
- * DualQuaternionSkinning applies a set of dual quaternions to a given input set of vertices and
- * returns the resulting transformed vertices.
+/**
+ * \brief Applies the given Dual-Quaternions to the given vertices.
+ */
```

Comment:
```suggestion
 * \warning Parallelized loop inside (using openmp), openmp cannot be used by the caller.
 */
```

Reply:
Well, support for parallel nested loop is available on OMP but is not active by default. According to the spec of OpenMP, on nested parallel loop, only the englobing loop is parallelized except if the (deprecated in openmp 5.0) OMP_NESTED  https://www.openmp.org/spec-html/5.1/openmpse66.html#x332-5070006.9 variable is set. So here, the doc can be informational but no need to set a warning as thereis no warning. If the caller is inside a parallel lopp, the inner loop is not parallelized and this change nothing for the caller state.

Reply:
Does this work if the nested loop is in a library, and the caller dynamically links with this library ?

Reply:
I think so. do you have a test for that ?

Reply:
Not really.
It’s just that I remember having problems with this long time ago, I guess openmp changed in the meantime. On windows maybe.